File: first_bad_version/first_bad_version.py
def isBadVersion(version):
    print("version: {}".format(version))
    if version >= 13:
        return True
    return False

def get_first_bad_version(n):
    left = 1
    right = n
    last_bad_version = -1
    while left < right:
        mid = (left + right) // 2        
        if isBadVersion(mid):           
            right = mid
        else:
            left = mid + 1
    return left

for n in (128, 64, 32, 16, 28, 24, 26):
    print("First bad of {}: {}".format(n, get_first_bad_version(n)))
    print("===================================================")

File: first_bad_version/first_bad_version.cpp
#include <vector>
#include <iostream>

static bool isBadVersion(int version)
{
    std::cout<<__func__<<"("<<version<<")\n";
    return version >= 13;
}

static int get_first_bad_version(int n)
{
    int left = 1;
    int right = n;
    while(left < right)
    {
        auto mid = (left + right) / 2;
        if (isBadVersion(mid))
        {
            right = mid;
        }
        else
        {
            left = mid + 1;
        }
    }

    return left;
}


int main(void)
{
    for(auto n: std::vector{128, 64, 32, 16, 28, 24, 26})
    {
        auto bad = get_first_bad_version(n);
        std::cout<<"Bad version of "<<n<<": "<<bad<<"\n";
        std::cout<<"=========================================\n";
    }
}